src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+#include "igraph_paths.h"
+#include "igraph_adjlist.h"
+#include "igraph_types.h"
+#include "igraph_memory.h"
+#include "igraph_error.h"
+#include "igraph.h"
+#include "igraph_interface.h"
+
+#include <cstring>
+#include <cmath>
+#include <map>
+#include <climits>
+#include <vector>
+#include <set>
+#include <algorithm>
+#include <iterator>
+
+/*
+ *
+ * Usage: Data structure to store index value of subsets.
+ * The index would be used in DP table
+ *
+ */
+std::map<std::set<igraph_integer_t>, igraph_integer_t> subsetMap;
```

Comment:
Using global variables like this is problematic. What if igraph is used from multiple threads? That could be fixed by adding `IGRAPH_THREAD_LOCAL`, but there are other problems too. What if an error occurs in the funtcion, or the computation is interrupted? In such cases memory should be released, and in C++ the best way to do it is with RAII. Ideally this should be refactored to be a local variable, possibly in the main Steiner tree function.